kythe/go/localrun/localrun.go

```diff
@@ -0,0 +1,503 @@
+		}
+
+		action := event.GetAction()
+		if action == nil {
+			continue
+		}
+
+		if !action.Success {
+			continue
+		}
+		if !r.Languages.hasExtractor(action.Type) {
+			fmt.Sprintf("Didn't find extractor for: %s\n", action.Type)
+			continue
+		}
+
+		actionCompleted := event.Id.GetActionCompleted()
+		if actionCompleted == nil {
+			continue
+		}
+
+		kzips = append(kzips, actionCompleted.PrimaryOutput)
+	}
+
+	fmt.Fprintf(os.Stderr, "Finished finding indexable targets\n")
+
+	indexer := &serialIndexer{
+		besFile:      r.besFile,
+		Languages:    r.Languages,
+		KytheRelease: r.KytheRelease,
+		WorkingDir:   r.WorkingDir,
+		Timeout:      r.Timeout,
+	}
+
+	r.indexedOut, err = indexer.run(ctx, kzips)
+	return err
+}
+
+// PostProcess postprocesses the supplied indexed data.
+func (r *Runner) PostProcess(ctx context.Context) error {
+	if err := r.checkMode(postprocess); err != nil {
+		return err
+	}
+
+	fmt.Fprintf(os.Stderr, "\nStarting postprocessing\n")
+
+	rd, err := dedup.NewReader(r.indexedOut, int(r.CacheSize.Bytes()))
+	if err != nil {
+		return fmt.Errorf("error creating deduped stream: %v", err)
+	}
+
+	tmpfile, err := ioutil.TempFile("", "dedup_kythe_stream.pb")
+	if err != nil {
+		return fmt.Errorf("error creating tmpfile: %v", err)
+	}
+	defer os.Remove(tmpfile.Name())
+
+	wr := delimited.NewWriter(tmpfile)
+	if err := delimited.Copy(wr, rd); err != nil {
+		return fmt.Errorf("unable to copy deduped stream to disk: %v", err)
+	}
+
+	p, s := beam.NewPipelineWithRoot()
+	entries, err := beamio.ReadEntries(ctx, s, tmpfile.Name())
+	if err != nil {
+		return fmt.Errorf("error reading entries: %v", err)
+	}
+	k := pipeline.FromEntries(s, entries)
+
+	beamio.WriteLevelDB(s, r.OutputDir, r.WorkerPoolSize,
+		createColumnarMetadata(s),
+		k.SplitCrossReferences(),
+		k.SplitDecorations(),
+		k.CorpusRoots(),
+		k.Directories(),
+		k.Documents(),
+		k.SplitEdges(),
+	)
+
+	return beamx.Run(ctx, p)
+
+	// dedup the stream
+	// https://github.com/kythe/kythe/blob/77e1360beca7903bc97728097d2c5e1b82204092/kythe/go/platform/tools/dedup_stream/dedup_stream.go
+	// https://github.com/kythe/kythe/blob/77e1360beca7903bc97728097d2c5e1b82204092/kythe/go/serving/tools/write_tables/write_tables.go#L139
+}
+
+func createColumnarMetadata(s beam.Scope) beam.PCollection {
+	return beam.ParDo(s, emitColumnarMetadata, beam.Impulse(s))
+}
+
+func emitColumnarMetadata(_ []byte) (string, string) { return xrefs.ColumnarTableKeyMarker, "v1" }
+
+// Serve starts an http server on the provided port and allows web inspection
+// of the indexed data.
+func (r *Runner) Serve(ctx context.Context) error {
+	if err := r.checkMode(serve); err != nil {
+		return err
+	}
+
+	listen := fmt.Sprintf(":%d", r.Port)
+
+	// Perform the work prescribed
+	cmd := exec.CommandContext(ctx,
+		fmt.Sprintf("%s/tools/http_server", r.KytheRelease),
+		"--listen", listen,
+		"--serving_table", r.OutputDir,
+		"--public_resources", fmt.Sprintf("%s/resources/public", r.KytheRelease),
```

Comment:
The web UI is no longer in the release archive.

Reply:
What would you recommend I put in here then?

Reply:
Unfortunately, we have no replacement currently.  Maybe do a check for the directory and add the `--public_resources` flag if it's found.  That way, it'll at least work with older releases.